### ISSUE TITLE
Panic on "Request to format inverted span"

### ISSUE THREAD

**Issue by gwenn** [rust-lang/rustfmt#3429]
Compiler is fine:
```sh
$ cargo check
    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
```

```sh
$ cargo fmt --version
rustfmt 1.0.3-nightly (d6829d62 2019-02-14)
```

```sh
$ RUST_BACKTRACE=1 cargo fmt
thread 'main' panicked at 'Request to format inverted span: Loc { file: SourceFile(/Users/gwen/Rust/lemon-rs/parser/src/parse.rs), line: 5559, col: CharPos(70), col_display: 70 } to Loc { file: SourceFile(/Users/gwen/Rust/lemon-rs/parser/src/parse.rs), line: 5559, col: CharPos(69), col_display: 69 }', src/tools/rustfmt/src/missed_spans.rs:97:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::_print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::continue_panic_fmt
   6: std::panicking::begin_panic_fmt
   7: rustfmt_nightly::missed_spans::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::format_missing_with_indent
   8: rustfmt_nightly::visitor::FmtVisitor::visit_block
   9: rustfmt_nightly::expr::rewrite_block_with_visitor
  10: rustfmt_nightly::expr::format_expr
  11: rustfmt_nightly::expr::<impl rustfmt_nightly::rewrite::Rewrite for syntax::ast::Stmt>::rewrite
  12: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  13: rustfmt_nightly::visitor::FmtVisitor::visit_block
  14: rustfmt_nightly::expr::rewrite_block_with_visitor
  15: rustfmt_nightly::expr::format_expr
  16: <rustfmt_nightly::matches::ArmWrapper<'a> as rustfmt_nightly::rewrite::Rewrite>::rewrite
  17: <rustfmt_nightly::lists::ListItems<'a, I, F1, F2, F3> as core::iter::traits::iterator::Iterator>::next
  18: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter
  19: rustfmt_nightly::matches::rewrite_match
  20: rustfmt_nightly::expr::format_expr
  21: rustfmt_nightly::expr::<impl rustfmt_nightly::rewrite::Rewrite for syntax::ast::Stmt>::rewrite
  22: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  23: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  24: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  25: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  26: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  27: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  28: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  29: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  30: rustfmt_nightly::visitor::FmtVisitor::walk_stmts
  31: rustfmt_nightly::visitor::FmtVisitor::visit_block
  32: rustfmt_nightly::visitor::FmtVisitor::visit_fn
  33: rustfmt_nightly::visitor::FmtVisitor::visit_impl_item
  34: rustfmt_nightly::items::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::visit_impl_items
  35: rustfmt_nightly::items::format_impl
  36: rustfmt_nightly::visitor::FmtVisitor::visit_item
  37: rustfmt_nightly::reorder::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::visit_items_with_reordering
  38: rustfmt_nightly::visitor::FmtVisitor::walk_mod_items
  39: rustfmt_nightly::visitor::FmtVisitor::format_separate_mod
  40: rustfmt_nightly::formatting::format_project
  41: syntax::with_globals
  42: <rustfmt_nightly::Session<'b, T>>::format
  43: rustfmt::format_and_emit_report
  44: <rustfmt_nightly::Session<'b, T>>::override_config
  45: rustfmt::execute
  46: rustfmt::main
  47: std::rt::lang_start::{{closure}}
  48: std::panicking::try::do_call
  49: __rust_maybe_catch_panic
  50: std::rt::lang_start_internal
  51: main
```

Source code looks like:
```rust
      306 /* vtabarglist ::= vtabarg */
     => {
//line 1249 "./parser/src/parse.y"
{yylhsminor = YYMINORTYPE::yy34( vec![self.yy_move(0).yy0().unwrap()]}
  self[0].minor = yylhsminor;
        }
```
Panic can be fixed by adding a semicolon:
```rust
{yylhsminor = YYMINORTYPE::yy34( vec![self.yy_move(0).yy0().unwrap()];}
```

Sources are [here](https://github.com/gwenn/lemon-rs).
`lemon-rs/parser/src/parse.rs` is generated by lemon so you need to run `make` before `rustfmt`.
Regards.

**Comment by topecongiro** [comment]
Looks like the opening paren of `yy34` does not have the corresponding closing paren, which is confusing rustfmt. I am not sure why this code is able to compile, though.

**Comment by gwenn** [comment]
Indeed, there is a missing closing parenthesis !

**Comment by gshowell** [comment]
I hit the same, also with a missing parenthesis. Same rustfmt version.

```rust
    #[test]
    fn test_add_piece() {
        ...
        assert!(empty.add_piece(Position::new(A1, Color::Black, Piece::King)).unwrap().add_piece(Position::new(A8, Color::Black, Piece::King)).is_ok();
    }
```

```sh
$ RUST_BACKTRACE=1 cargo fmt
thread 'main' panicked at 'Request to format inverted span: Loc { file: SourceFile(/home/falafel/checkouts/chess/src/board/squareboard.rs), line: 552, col: CharPos(5), col_display: 5 } to Loc { file: SourceFile(/home/falafel/checkouts/chess/src/board/squareboard.rs), line: 552, col: CharPos(4), col_display: 4 }', src/tools/rustfmt/src/missed_spans.rs:97:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:336
   7: rustfmt_nightly::missed_spans::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::format_missing_with_indent
   8: rustfmt_nightly::visitor::FmtVisitor::visit_block
   9: rustfmt_nightly::visitor::FmtVisitor::visit_fn
  10: rustfmt_nightly::visitor::FmtVisitor::visit_item
  11: rustfmt_nightly::reorder::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::visit_items_with_reordering
  12: rustfmt_nightly::visitor::FmtVisitor::walk_mod_items
  13: rustfmt_nightly::visitor::FmtVisitor::visit_item
  14: rustfmt_nightly::reorder::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::visit_items_with_reordering
  15: rustfmt_nightly::visitor::FmtVisitor::walk_mod_items
  16: rustfmt_nightly::visitor::FmtVisitor::format_separate_mod
  17: rustfmt_nightly::formatting::format_project
  18: <scoped_tls::ScopedKey<T>>::set
  19: syntax::with_globals
  20: <rustfmt_nightly::Session<'b, T>>::format
  21: rustfmt::format_and_emit_report
  22: <rustfmt_nightly::Session<'b, T>>::override_config
  23: rustfmt::execute
  24: rustfmt::main
  25: std::rt::lang_start::{{closure}}
  26: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  27: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  28: std::rt::lang_start_internal
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  29: main
  30: __libc_start_main
  31: <unknown>
```

Switching to stable and running `rustfmt 1.0.1-stable (be13559 2018-12-10)` correctly errors

```sh
$ cargo fmt
error: incorrect close delimiter: `}`
   --> /home/falafel/checkouts/chess/src/board/squareboard.rs:552:5
    |
531 |     fn test_add_piece() {
    |                         - close delimiter possibly meant for this
...
551 |         assert!(empty.add_piece(Position::new(A1, Color::Black, Piece::King)).unwrap().add_piece(Position::new(A8, Color::Black, Piece::King)).is_ok();
    |                - un-closed delimiter
552 |     }
    |     ^ incorrect close delimiter

Error writing files: parse error
```

**Comment by topecongiro** [comment]
This is fixed after we bumped `rustc-ap-syntax` to 407.0.0, so it was probably due to the rustc parser error.
